### PR TITLE
2759

### PR DESCRIPTION
**Pull request by assaferan**
Resolving issue #2759.

### CODE REVIEW

File: lmfdb/classical_modular_forms/web_newform.py [hunk 1]
```diff
@@ -954,6 +954,55 @@ def td_wrap(val):
...
+      # replaced by a call to a single class/function with some parameters
+        def th_wrap(kwl, title):
+            return '    <th>%s</th>' % display_knowl(kwl, title=title)
+        def td_wrap(val):
+            return '    <td>$%s$</th>' % val
+        data = db.mf_newforms.lookup(self.label)
+        hecke_orbit_code = data['hecke_orbit_code']
+        self.heckepolys = []
+        R = PolynomialRing(ZZ, 'T')
+        T = R.gen(0)
```
Comment (edgarcosta) [inline lmfdb/classical_modular_forms/web_newform.py]: This line needs to be deleted as the variable `T` is never used.

Comment (assaferan) [reply lmfdb/classical_modular_forms/web_newform.py]: Oops, sorry. Also noticed that I should add the change to new spaces as well. On it.

**Comment by assaferan** [comment]
Ok, I think that now it's good to go - it displays the polynomials for newform spaces (also for Gamma1 spaces) and for the newforms themselves. Also, I got rid of the unused variable.

**Comment by AndrewVSutherland** [comment]
The travis failure is pylint complaining about line 197 of web_space.py:

web_space.py:    heckepolys = hecke_polys_orbits.items()

The variable heckepolys is used in two places below

web_space.py:    if len(heckepolys) == 0:
web_space.py:    for p, lpoly in heckepolys:

While this warning is pretty mild, and I think the code you wrote is valid, rather than disable the warning I think it would be better to change your code to avoid it.

You can replace "if len(heckepolys) == 0" with "if not hecke_polys_orbits" (which is arguable better in any case), and then replace heckepolys with hecke_polys_orbits.items() in the for statement in line 295.

**Comment by assaferan** [comment]
Thanks!

**Comment by AndrewVSutherland** [comment]
So this works and I like how it looks, but there is a performance issue that we need to deal with before this PR can be merged.  In high weight cases where the Hecke polys have huge coefficients the pages take a very long time to load, e.g.

http://127.0.0.1:37777/ModularForm/GL2/Q/holomorphic/1/254/a/a/

takes about 25 seconds, vs less than 1 second for

https://www.lmfdb.org/ModularForm/GL2/Q/holomorphic/1/254/a/a/

This is an extreme example (we don't even try to compute the rational L-function for newforms with weight over 200),  but a slightly less extreme example is

http://127.0.0.1:37777/ModularForm/GL2/Q/holomorphic/1/198/a/a/

but you can see the same L-polynomials much more quickly in a condensed display that uses bigint knowls on the rational L-function page:

https://www.lmfdb.org/L/ModularForm/GL2/Q/holomorphic/1/198/a/a/

Now this page is still extracting just as much data from the database and sending just as much data to the browser (more in fact), so the code that is generating the Hecke poly must be less efficient in some way.

Before we can merge this we need to (1) suppress the display in cases where the charpolys are simply too big to load in a reasonable amount of time (or maybe just give the first 5 rather than 25 of them), (2) figure out why it is taking so much longer to construct and display this table than it is on the rational L-function pages, and (3) if we don't decide to suppress display of L-polynomials with large coefficients, we should use bigint knowls to keep the amount of horizontal scrowling reasonable (and possibly this also speeds things up and addresses (2), it would be wroth trying in any case).

@edgarcosta do you have any thoughts on this?

**Comment by assaferan** [comment]
Thanks Drew - meanwhile I added the displaying of the coefficients as bigints.
After profiling it seems that the factoring is taking most of the time (in the above example http://127.0.0.1:37777/ModularForm/GL2/Q/holomorphic/1/198/a/a/ - it takes 6.5 out of the 7 seconds)
I noticed that in the L functions data the Euler factors are already factored, and so the server doesn't need to do any of this work. Is there a reason we only have it for the L-function access and not for the mf_hecke_lpoly ?

Also, in the L-functions page, I saw that the scrolling is inside the table (and not all over the page) - I'll try to do it here as well.

**Comment by edgarcosta** [comment]
The reason is exactly what @assaferan already found out, factoring takes a long time.
The clean solution is to add the factorization to mf_hecke_lpoly, or to only store the factorization.
The hacky solutions is to grab the factorization from lfunc_lfunctions.

**Comment by assaferan** [comment]
How do I add things to mf_hecke_lpoly, then?

**Comment by AndrewVSutherland** [comment]
It probably makes sense for me to add the factorizations -- the mf_hecke_lpoly table is populated from an external file created by an external program (FormatNewformData in https://github.com/JohnCremona/CMFs/blob/master/magma/mfpgdata.m) that generates all of the newform tables from (many gigabytes worth) of raw data files.

It would not be hard for me to update this program to also compute factorizations of the hecke lpolys and include them as a column in mf_hecke_lpoly table, and rerun the table generation script (it only takes about 3 hours running on 64 cores), but this will need to wait until next week.

**Comment by AndrewVSutherland** [comment]
The table mf_hecke_lpolys now contains the column lpoly_factorization, which contains the factorization of the L-polynomial into irreducible factors as [[[list-of-coeffs],exponent],...]. See

https://beta.lmfdb.org/api/mf_hecke_lpolys/

Every factor should have constant coefficient 1, and they should be sorted by degree, then exponent, then by coefficients (lex order).

**Comment by assaferan** [comment]
OK - the code now retrieves the factored polynomials and works much faster. Do you think it's good to go?

**Comment by AndrewVSutherland** [comment]
Looks good!

**Pull request merged**